File: src/main/java/minic/ui/MiniCSemanticScopeTreeLine.java
package minic.ui;

import java.util.List;
import java.util.Objects;

/**
 * Semantic 作用域树视图中的一行。
 *
 * @param label 作用域标签
 * @param depth 树深度
 * @param symbols 符号摘要
 * @param active 是否高亮
 * @param onActivePath 是否位于当前路径
 * @param arrowDirection 边方向
 */
public record MiniCSemanticScopeTreeLine(
        String label,
        int depth,
        List<String> symbols,
        boolean active,
        boolean onActivePath,
        String arrowDirection
) {
    public MiniCSemanticScopeTreeLine {
        Objects.requireNonNull(label, "label");
        Objects.requireNonNull(symbols, "symbols");
        Objects.requireNonNull(arrowDirection, "arrowDirection");
        if (depth < 0) {
            throw new IllegalArgumentException("depth must not be negative");
        }
        symbols = List.copyOf(symbols);
    }
}

File: uiweb/src/visual/MiniCSemanticScopeTreeLine.ts
import type { JavaMirrorFile } from "../translation/javaMirror";

export const miniCSemanticScopeTreeLineMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCSemanticScopeTreeLine.java",
  webPath: "uiweb/src/visual/MiniCSemanticScopeTreeLine.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCSemanticScopeTreeLine",
  kind: "record",
  imports: ["java.util.List", "java.util.Objects"],
  fields: [],
  methods: [{ name: "MiniCSemanticScopeTreeLine", signature: "MiniCSemanticScopeTreeLine(String label, int depth, List<String> symbols, boolean active, boolean onActivePath, String arrowDirection)" }],
} as const satisfies JavaMirrorFile;

export type MiniCSemanticScopeArrowDirection = "child-to-parent" | "parent-to-child";

export class MiniCSemanticScopeTreeLine {
  static readonly mirror = miniCSemanticScopeTreeLineMirror;

  readonly symbols: readonly string[];

  constructor(
    readonly label: string,
    readonly depth: number,
    symbols: readonly string[],
    readonly active: boolean,
    readonly onActivePath: boolean,
    readonly arrowDirection: MiniCSemanticScopeArrowDirection,
  ) {
    if (depth < 0) {
      throw new Error("depth must not be negative");
    }
    this.symbols = [...symbols];
  }
}

export default MiniCSemanticScopeTreeLine;
